File: msbuild/Fspure.BuildTasks/EmbedPureJson.fs
namespace Fspure.BuildTasks

open System
open System.IO
open Microsoft.Build.Framework
open Microsoft.Build.Utilities
open Mono.Cecil

/// Injects (or replaces) an embedded pure.json resource into an already-built managed assembly.
/// Used after fspure-collector runs against $(TargetPath).
type EmbedPureJson() =
    inherit Task()

    /// Path to the built assembly (DLL/EXE) to mutate in place.
    [<Required>]
    member val AssemblyPath = "" with get, set

    /// Path to the .pure.json file to embed.
    [<Required>]
    member val PureJsonPath = "" with get, set

    /// Manifest resource name. Convention: {AssemblyName}.pure.json.
    [<Required>]
    member val ResourceName = "" with get, set

    static member private TryAddPackageLibs(resolver: DefaultAssemblyResolver, nugetRoot: string, packageId: string) =
        try
            let pkg = Path.Combine(nugetRoot, packageId)

            if Directory.Exists pkg then
                for verDir in Directory.GetDirectories pkg |> Array.sortDescending do
                    let lib = Path.Combine(verDir, "lib")

                    if Directory.Exists lib then
                        for tfm in Directory.GetDirectories lib do
                            resolver.AddSearchDirectory tfm
        with _ ->
            ()

    static member private AddCommonSearchPaths(resolver: DefaultAssemblyResolver) =
        let nuget =
            match Environment.GetEnvironmentVariable "NUGET_PACKAGES" with
            | null
            | "" ->
                Path.Combine(
                    Environment.GetFolderPath Environment.SpecialFolder.UserProfile,
                    ".nuget",
                    "packages"
                )
            | p -> p

        EmbedPureJson.TryAddPackageLibs(resolver, nuget, "fsharp.core")

        let dotnetRoot =
            match Environment.GetEnvironmentVariable "DOTNET_ROOT" with
            | null
            | "" ->
                match Path.GetDirectoryName(Environment.ProcessPath) with
                | null
                | "" -> "/usr/share/dotnet"
                | d -> d
            | p -> p

        let shared = Path.Combine(dotnetRoot, "shared", "Microsoft.NETCore.App")

        if Directory.Exists shared then
            match Directory.GetDirectories shared |> Array.sortDescending |> Array.tryHead with
            | Some ver -> resolver.AddSearchDirectory ver
            | None -> ()

    override this.Execute() =
        try
            if String.IsNullOrWhiteSpace this.AssemblyPath || not (File.Exists this.AssemblyPath) then
                this.Log.LogError("Fspure EmbedPureJson: assembly not found: {0}", this.AssemblyPath)
                false
            elif String.IsNullOrWhiteSpace this.PureJsonPath || not (File.Exists this.PureJsonPath) then
                this.Log.LogError("Fspure EmbedPureJson: pure.json not found: {0}", this.PureJsonPath)
                false
            elif String.IsNullOrWhiteSpace this.ResourceName then
                this.Log.LogError("Fspure EmbedPureJson: ResourceName is required.")
                false
            else
                let jsonBytes = File.ReadAllBytes this.PureJsonPath
                let assemblyFull = Path.GetFullPath this.AssemblyPath

                let directory =
                    match Path.GetDirectoryName assemblyFull with
                    | null
                    | "" -> "."
                    | d -> d

                use resolver = new DefaultAssemblyResolver()
                resolver.AddSearchDirectory directory
                EmbedPureJson.AddCommonSearchPaths resolver

                let readerParameters =
                    ReaderParameters(
                        AssemblyResolver = resolver,
                        ReadWrite = false,
                        InMemory = true,
                        ReadingMode = ReadingMode.Immediate
                    )

                use moduleDef = ModuleDefinition.ReadModule(assemblyFull, readerParameters)

                for i = moduleDef.Resources.Count - 1 downto 0 do
                    if
                        String.Equals(moduleDef.Resources[i].Name, this.ResourceName, StringComparison.Ordinal)
                    then
                        moduleDef.Resources.RemoveAt i

                moduleDef.Resources.Add(
                    EmbeddedResource(this.ResourceName, ManifestResourceAttributes.Public, jsonBytes)
                )

                let tempPath = assemblyFull + ".fspure-tmp"

                try
                    moduleDef.Write tempPath
                    File.Copy(tempPath, assemblyFull, true)
                finally
                    if File.Exists tempPath then
                        try
                            File.Delete tempPath
                        with _ ->
                            ()

                this.Log.LogMessage(
                    MessageImportance.Low,
                    "Fspure: embedded resource '{0}' ({1} bytes) into {2}",
                    this.ResourceName,
                    jsonBytes.Length,
                    assemblyFull
                )

                true
        with ex ->
            this.Log.LogError("Fspure EmbedPureJson failed: {0}", ex.Message)
            this.Log.LogMessage(MessageImportance.Low, ex.ToString())
            false

File: msbuild/Fspure.BuildTasks/EmbedPureJson.cs
using Microsoft.Build.Framework;
using Microsoft.Build.Utilities;
using Mono.Cecil;

namespace Fspure.BuildTasks;

/// <summary>
/// Injects (or replaces) an embedded pure.json resource into an already-built managed assembly.
/// Used after fspure-collector runs against $(TargetPath).
/// </summary>
public sealed class EmbedPureJson : Microsoft.Build.Utilities.Task
{
    /// <summary>Path to the built assembly (DLL/EXE) to mutate in place.</summary>
    [Required]
    public string AssemblyPath { get; set; } = "";

    /// <summary>Path to the .pure.json file to embed.</summary>
    [Required]
    public string PureJsonPath { get; set; } = "";

    /// <summary>
    /// Manifest resource name. Convention: <c>{AssemblyName}.pure.json</c>.
    /// </summary>
    [Required]
    public string ResourceName { get; set; } = "";

    public override bool Execute()
    {
        try
        {
            if (string.IsNullOrWhiteSpace(AssemblyPath) || !File.Exists(AssemblyPath))
            {
                Log.LogError("Fspure EmbedPureJson: assembly not found: {0}", AssemblyPath);
                return false;
            }

            if (string.IsNullOrWhiteSpace(PureJsonPath) || !File.Exists(PureJsonPath))
            {
                Log.LogError("Fspure EmbedPureJson: pure.json not found: {0}", PureJsonPath);
                return false;
            }

            if (string.IsNullOrWhiteSpace(ResourceName))
            {
                Log.LogError("Fspure EmbedPureJson: ResourceName is required.");
                return false;
            }

            var jsonBytes = File.ReadAllBytes(PureJsonPath);
            var assemblyFull = Path.GetFullPath(AssemblyPath);
            var directory = Path.GetDirectoryName(assemblyFull) ?? ".";

            var resolver = new DefaultAssemblyResolver();
            resolver.AddSearchDirectory(directory);
            AddCommonSearchPaths(resolver);

            var readerParameters = new ReaderParameters
            {
                AssemblyResolver = resolver,
                ReadWrite = false,
                InMemory = true,
                ReadingMode = ReadingMode.Immediate,
            };

            using var module = ModuleDefinition.ReadModule(assemblyFull, readerParameters);

            // Remove any prior pure.json resource with the same logical name.
            for (var i = module.Resources.Count - 1; i >= 0; i--)
            {
                if (string.Equals(module.Resources[i].Name, ResourceName, StringComparison.Ordinal))
                {
                    module.Resources.RemoveAt(i);
                }
            }

            module.Resources.Add(
                new EmbeddedResource(ResourceName, ManifestResourceAttributes.Public, jsonBytes));

            var tempPath = assemblyFull + ".fspure-tmp";
            try
            {
                module.Write(tempPath);
                File.Copy(tempPath, assemblyFull, overwrite: true);
            }
            finally
            {
                if (File.Exists(tempPath))
                {
                    try { File.Delete(tempPath); } catch { /* ignore */ }
                }
            }

            Log.LogMessage(
                MessageImportance.Low,
                "Fspure: embedded resource '{0}' ({1} bytes) into {2}",
                ResourceName,
                jsonBytes.Length,
                assemblyFull);

            return true;
        }
        catch (Exception ex)
        {
            Log.LogError("Fspure EmbedPureJson failed: {0}", ex.Message);
            Log.LogMessage(MessageImportance.Low, ex.ToString());
            return false;
        }
    }

    private static void AddCommonSearchPaths(DefaultAssemblyResolver resolver)
    {
        // NuGet global packages (FSharp.Core, etc.)
        var nuget =
            Environment.GetEnvironmentVariable("NUGET_PACKAGES")
            ?? Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
                ".nuget",
                "packages");

        TryAddPackageLibs(resolver, nuget, "fsharp.core");

        // .NET shared framework (System.* facades)
        var dotnetRoot =
            Environment.GetEnvironmentVariable("DOTNET_ROOT")
            ?? Path.GetDirectoryName(Environment.ProcessPath)
            ?? "/usr/share/dotnet";

        var shared = Path.Combine(dotnetRoot, "shared", "Microsoft.NETCore.App");
        if (Directory.Exists(shared))
        {
            foreach (var ver in Directory.GetDirectories(shared).OrderByDescending(d => d))
            {
                resolver.AddSearchDirectory(ver);
                break;
            }
        }
    }

    private static void TryAddPackageLibs(DefaultAssemblyResolver resolver, string nugetRoot, string packageId)
    {
        try
        {
            var pkg = Path.Combine(nugetRoot, packageId);
            if (!Directory.Exists(pkg))
            {
                return;
            }

            foreach (var verDir in Directory.GetDirectories(pkg).OrderByDescending(d => d))
            {
                var lib = Path.Combine(verDir, "lib");
                if (!Directory.Exists(lib))
                {
                    continue;
                }

                foreach (var tfm in Directory.GetDirectories(lib))
                {
                    resolver.AddSearchDirectory(tfm);
                }
            }
        }
        catch
        {
            // Best-effort only.
        }
    }
}
